File: programmers.co.kr/complete_search/2.cpp
// https://programmers.co.kr/learn/courses/30/lessons/42839
// Finding Primes

#include <cmath>
#include <string>
#include <unordered_set>
#include <vector>

using namespace std;

// pre: num is a nonnegative integer
bool isPrime(int num) {
    if (num == 2) {
        return true;
    } else if (num == 0 || num == 1 || num % 2 == 0) {
        return false;
    }

    for (int i = 3; i <= sqrt(num); i += 2) {
        if (num % i == 0) {
            return false;
        }
    }
    return true;
}

int searchAll(const string &numbers, string &builder, int index,
              unordered_set<int> &seen) {
    if (index == numbers.length()) {
        if (builder.empty() || seen.find(stoi(builder)) != seen.end()) {
            return 0;
        }
        int check = stoi(builder);
        seen.insert(check);
        return isPrime(check);
    }
    // Generate subset with index num excluded
    int result = searchAll(numbers, builder, index + 1, seen);
    // Generate subset with index num included, all permutations
    for (int i = 0; i <= builder.length(); i++) {
        char toInsert = numbers[index];
        builder = builder.substr(0, i) + toInsert + builder.substr(i);
        result += searchAll(numbers, builder, index + 1, seen);
        builder = builder.substr(0, i) + builder.substr(i + 1);
    }
    return result;
}

int solution(string numbers) {
    string builder = "";
    unordered_set<int> seen;
    int answer = searchAll(numbers, builder, 0, seen);
    return answer;
}

File: programmers.co.kr/complete_search/2.py
# https://programmers.co.kr/learn/courses/30/lessons/42839
# Finding Primes

from math import floor, sqrt

class Searcher:
    seen = set()

    def __init__(self, numbers):
        self.numbers = numbers

    def search(self, builder, level):
        if (level == len(self.numbers)):
            if not builder or int(builder) in self.seen:
                return 0
            self.seen.add(int(builder))
            return int(isPrime(int(builder)))

        else:
            # Generate subset excluding current number
            count = self.search(builder, level+1)
            # Generate subset including current number, all permutations
            for i in range(len(builder)+1):
                builderPermute = builder[0:i] + self.numbers[level] + builder[i:len(builder)]
                count += self.search(builderPermute, level+1)
            return count


def isPrime(num):
    '''precondition: nums is a nonnegative integer.'''
    if (num == 2):
        return True
    elif (num == 0 or num == 1 or num % 2 == 0):
        return False

    for i in range(3, floor(sqrt(num))+1, 2):
        if (num % i == 0):
            return False
    return True


def solution(numbers):
    s = Searcher(numbers)
    return s.search("", 0)
